cpp/src/io/csv/csv_gpu.cu

```diff
@@ -570,6 +571,315 @@ __global__ void convertCsvToGdf(const char *raw_csv,
   }
 }
 
+/**
+ * @brief merge two packed row contexts (each corresponding to a block of characters)
+ * and returns the packed row context corresponding to the merged character block
+ **/
+inline __device__ uint64_t merge_row_contexts(uint64_t first_ctx, uint64_t second_ctx)
+{
+  uint32_t id0 = get_row_context(first_ctx, ROW_CTX_NONE) & 3;
+  uint32_t id1 = get_row_context(first_ctx, ROW_CTX_QUOTE) & 3;
+  uint32_t id2 = get_row_context(first_ctx, ROW_CTX_COMMENT) & 3;
+  return (first_ctx & ~pack_row_contexts(3, 3, 3)) +
+         pack_row_contexts(get_row_context(second_ctx, id0),
+                           get_row_context(second_ctx, id1),
+                           get_row_context(second_ctx, id2));
+}
+
+/**
+ * @brief Short row context:
+ * 1-bit count (0 or 1) per context in the lower 4 bits
+ * 2-bit output context id per input context in bits 8..15
+ **/
+inline __device__ uint32_t make_short_row_context(uint32_t id0,
+                                                  uint32_t id1,
+                                                  uint32_t id2 = ROW_CTX_COMMENT,
+                                                  uint32_t c0  = 0,
+                                                  uint32_t c1  = 0,
+                                                  uint32_t c2  = 0)
+{
+  return (id0 << 8) | (id1 << 10) | (id2 << 12) | (ROW_CTX_EOF << 14) | (c0) | (c1 << 1) |
```

Comment:
There's a lot of hard-coded constants. It would be great if they were named.
IMO it would be best to declare field sizes as constexpr ints, and use these constants in context struct declaration(s), and also in the implementation.
Anything derived from these sizes (masks, offsets) can be also computed at compile time so the final code will effectively be the same as the hard-coded version.

Reply:
IMO I actually find this more readable (along with the comment immediately above), rather than defining something like:
```
enum { BIT_POS_ID0 = 8 + 0*2, BIT_POS_ID1 = 8 + 1*2, BIT_POS_ID2 = 8 + 2*2, BIT_POS_ID3 = 8 + 3 * 2 };
return (id0 << BIT_POS_ID0) | (id1 << BIT_POS_ID1) etc..
```
It really doesn't matter what the values of the constants are as long as it matches between `make_short_row_context` and `merge_short_row_context` (the only two places where these constants are used), so in that sense, I don't think it detracts from understanding the merging operation.

Reply:
I think the comment two lines above makes it clearer than a bunch of constexpr values defined further away:
```
 * @brief Short row context:
 * 1-bit count (0 or 1) per context in the lower 4 bits
 * 2-bit output context id per input context in bits 8..15
```

Reply:
I'll focus on adding more comments instead.